File: Cpp/wordle_simu.cpp
#include <vector>
#include <iostream>
#include <string>
#include <algorithm>
#include "read_csv.hpp"
#include "help_functions.hpp"
#include "wordle_game.hpp"

double worlde_simu(std::vector<std::string> & red_pp, std::vector<std::string> &allowed_guesses, std::vector<std::string> &allowed_answers,unsigned int allowed_attempts,std::string word, std::vector<std::string> &patterns)
{
    // Setup for each iteration
    
    red_pp = allowed_guesses;
    double attempt = 1.0;
    std:: string solution;
    solution = get_random(allowed_answers);
    std:: string pattern;
    std:: vector<double> exp_inf;
    int ind_max_exp_inf;
    std:: string best_word;

    
    while (attempt < allowed_attempts + 1){
        pattern = give_pattern(word,solution);

        red_pp = possible_words(word, pattern, red_pp);
        exp_inf = generate_exp_inf(red_pp,patterns,red_pp);

        ind_max_exp_inf = max_element(exp_inf.begin(),exp_inf.end()) - exp_inf.begin();
        best_word = red_pp[ind_max_exp_inf];

        if (word == solution){
            return attempt;
        }
        else {
            attempt++;
            word = best_word;
        }
    }
    return attempt;
}

void run_simu(std::vector<std::string> & red_pp,std::vector<std::string> &allowed_guesses, std::vector<std::string> &allowed_answers,unsigned iterations){
    double iter_divider = iterations + 0.0;
    double total_attempts = 0.0;
    double current_attempt;
    double avg_attempts;
    
    std::vector<std::string> patterns = generate_patterns();
    for (unsigned i = 0; i<iterations; i++){
        current_attempt = worlde_simu(red_pp,allowed_guesses,allowed_answers,6, "tares",patterns);
        total_attempts += current_attempt;
        std:: cout << i << "\n";
    }
    avg_attempts = total_attempts / iter_divider;
    std:: cout << "Iterationen: " << iterations << "\n";
    std:: cout << "Durschnittlichen Attempts: " << avg_attempts;
    
}

int main()
{
    std::vector<std::string> allowed_guesses = read_csv("Data/wordle_guesses.txt");
    
    std::vector<std::string> allowed_answers = read_csv("Data/wordle_answers.txt");

    // join both txt in vectors
    allowed_guesses.insert(allowed_guesses.end(), allowed_answers.begin(), allowed_answers.end());
    std::sort(allowed_guesses.begin(), allowed_guesses.end());

    // init variables
    std::vector<std::string> red_pp;

    run_simu(red_pp,allowed_guesses,allowed_answers,1000);
}

File: wordle_simu.py
#%%
import wordle_solver
import wordle_main
import wordle_statistics
import copy
import numpy as np
import test_cy
import char_position_statistics

#%%
def give_pattern(guess,solution):

    if guess == solution:

            return [2,2,2,2,2]

    else:

        # Create array which will encode feedback for every letter in guess
        feedback = []

        for i, letter in enumerate(guess):

            # Case 1: Letter at correct position
            if letter == solution[i]:
                feedback.append(2)

            # Case 2: Letter in word but at a different position
            elif letter in solution:
                feedback.append(1)

            # Case 3: Letter not in word
            else:
                feedback.append(0)
        
        return feedback

def run_simu(iterations):
    total_attempts= 0
    for i in range(iterations):
        current_attempt = worlde_simu(6,allowed_guesses)
        total_attempts+=current_attempt
        print(i)
    avg_attempts = total_attempts/iterations
    print(f"Die durchschnittlichen Attempts über {iterations} Iterationen betrugen: {avg_attempts}") 

def worlde_simu(allowed_attempts, allowed_guesses):

    #phase_space = copy.deepcopy(allowed_guesses)
    attempt = 1
    word = 'cares'
    
    patterns = wordle_statistics.generate_patterns()
    allowed_guesses, solutions = wordle_main.read_word_files(
        solutions_path, guesses_path)
    solution = np.random.choice(solutions)
    while attempt < allowed_attempts + 1:
        pattern = give_pattern(word,solution)

        #char comparison algorithm
        """ red_pp = char_position_statistics.possible_words(word, pattern, allowed_guesses)
        exp_inf = char_position_statistics.determine_frequency_for_each_word(red_pp) """
       
        #compute solution space algortihm
        red_pp = test_cy.possible_words(word, pattern, allowed_guesses)
        exp_inf = test_cy.generate_exp_inf(red_pp,patterns,red_pp)

        ind_max_exp_inf = np.argmax(exp_inf)
        best_word = red_pp[ind_max_exp_inf]

        allowed_guesses = red_pp

        if word == solution:
            return attempt
        else:
            attempt += 1
            
            word = best_word

    return attempt


if __name__ == "__main__":
    solutions_path = "Data/wordle_answers.txt"
    guesses_path = "Data/wordle_guesses.txt"

    allowed_guesses, solutions = wordle_main.read_word_files(
        solutions_path, guesses_path)
    run_simu(1000)
    #worlde_simu(6, allowed_guesses)
    #print(give_pattern('cares','tares'))

# %%
